plankton/core/devices.py

```diff
@@ -0,0 +1,349 @@
+
+        return None
+
+    @property
+    def interfaces(self):
+        """
+        This property contains a map with protocols as keys and interface types as values.
+        The types are imported from the ``interfaces`` sub-module and from the device module
+        itself. If two interfaces with the same protocol are discovered, a RuntimeError is raiesed.
+        """
+        return self._interfaces
+
+    @property
+    def protocols(self):
+        """All available protocols for this device."""
+        return list(self.interfaces.keys())
+
+    @property
+    def default_protocol(self):
+        """In case only one protocol exists for the device, this is the default protocol."""
+        if len(self.protocols) == 1:
+            return self.protocols[0]
+
+        return None
+
+    @property
+    def setups(self):
+        """
+        A map with all available setups. Setups are imported from the ``setups`` dictionary
+        in a device module and from the ``setups`` sub-module. If no ``default``-setup exists,
+        one is created using the default_device_type. If there are several device types in
+        the module, the default setup must be provided explicitly.
+        """
+        return self._setups
+
+    def _create_device_instance(self, device_type, **kwargs):
+        if device_type not in self.device_types:
+            raise RuntimeError('Can not create instance of non-device type.')
+
+        return device_type(**kwargs)
+
+    def create_device(self, setup=None):
+        """
+        Creates a device object according to the provided setup. If no setup is provided,
+        the default setup is used. If the setup can't be found, a PlanktonException is raised.
+        This can also happen if the device type specified in the setup is invalid.
+
+        :param setup: Name of the setup from which to create device.
+        :return: Device object initialized according to the provided setup.
+        """
+        setup_name = setup if setup is not None else 'default'
+
+        if setup_name not in self.setups:
+            raise PlanktonException(
+                'Failed to find setup \'{}\' for device \'{}\'. '
+                'Available setups are:\n    {}'.format(
+                    setup, self.name, '\n    '.join(self.setups.keys())))
+
+        setup_data = self.setups[setup_name]
+        device_type = setup_data.get('device_type') or self.default_device_type
+
+        try:
+            return self._create_device_instance(
+                device_type, **setup_data.get('parameters', {}))
+        except RuntimeError:
+            raise PlanktonException(
+                'The setup \'{}\' you tried to load does not specify a valid device type, but the '
+                'device module \'{}\' provides multiple device types so that no meaningful '
+                'default can be deduced.'.format(setup_name, self.name))
+
+    def create_interface_type(self, protocol=None):
```

Comment:
As referenced in another comment, couldn't this be changed to:

```python
def create_interface(self, protocol, device, args):
    ...
```
I can't find any usage of the default protocol value (None is explicitly passed in when not specified on command line) or needing to get the type rather than an instance.